### ISSUE TITLE
Wrong return code instantiating Jenkins build associated with Pipeline BuildConfig

### ISSUE THREAD

**Issue by csrwng** [openshift/jenkins-sync-plugin#175]
When starting a Jenkins job that is associated with a Pipeline BuildConfig through the REST URL, the return code is a 302 which redirects to the Jenkins job page.

If a job is not associated with a Pipeline BuildConfig, Jenkins correctly returns a 201 with the URL to the newly created run.

To reproduce:

1. Create a new pipeline BuildConfig such as 
    https://github.com/openshift/origin/blob/master/examples/jenkins/pipeline/maven-pipeline.yaml
2. Obtain the current user's token to use as a bearer token:
    ```
    mytoken="$(oc whoami -t)"
    ````

3. Start the build using curl against the Jenkins server:
    ```
    curl -k -v -X POST -H "Authentication: Bearer ${mytoken}"  https://[jenkins-server-url]/job/[namespace]-maven-pipeline/buildWithParameters
    ```
Observe the result of the POST: It's going to be a 302 back to the Jenkins page for the job.

You can compare with a Job created directly on the Jenkins server that doesn't have a corresponding BuildConfig associated with it. The POST returns a 201 with a Location header indicating the build URL for the newly created run.

**Comment by csrwng** [comment]
@gabemontero

**Comment by bparees** [comment]
@csrwng was the job you created by hand also a workflow/pipeline job or was it a classic job?

**Comment by csrwng** [comment]
A workflow/pipeline job

**Comment by gabemontero** [comment]
@csrwng and I talked about this some.  Another point of comparison might be starting pipeline builds from the jenkins console vs. `oc start-build`.

Bottom line, my suspicion is that when massaging the Jenkins `Run` for pipeline builds, it is doing so in way that influences Jenkins to react different at the servlet/http layer when reporting back to the client.

When I get to this, I believe I can add some diagnostics in our login plugin servlet filter, as well as the start of our various plug points into Jenkins (including our listeners which should see the non-pipeline jobs (even though by default it ignores them), and start to decipher the differences.

Where hopefully we can reconcile those differences once we understand the specifics.

Or there is some jenkins induced limitation, and we can't support @csrwng 's scenario here.  But I'm "glass half full" still and think we can avoid that.

**Comment by gabemontero** [comment]
That difference in behavior might be waiting on the build object getting created in the "start the build via the jenkins console" before we finish our `Run` patching and returning to Jenkins during its various callouts to sync plugin.

**Comment by gabemontero** [comment]
OK, I've pieced a few things together.

1) the 302 and the location header @csrwng originally reported is occurring because of https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/security/BasicAuthenticationFilter.java#L95-L159

2) the question is why we hit the `HttpServletResponse.SC_MOVED_TEMPORARILY` portion of this path when triggering a pipeline job

3) jenkins tracing is minimal, and patching / replacing the jenkins war file and not breaking dependencies has been problematic in the past (though I might give it a go ... we'll see) ....

4) but I suspect it is related to the sync plugin scheduling the job, and replacing/overriding the scheduling that occurs at places like https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/model/ParameterizedJobMixIn.java#L216-L222 and similar locations that set the 201 and redirect @csrwng is looking for... those points in Jenkis are where I suspect Jenkins overrides the 302/location header manipulation done in BasicAuthenticationFilter.java.  These points with the 201, etc. are basically at the webapp level where the Jenkins console triggers the build/run.

5) this replace/override is unavoidable; it adds the metadata needed to link the job/run to the bc/bld

Looking like we might have to punt on this, but not ready to do that just yet.  Where I'm going to go next is start examining the Jenkins run queue at the various plugin insertion points into jenkins in the sync plugin.  If my speculation is correct, going to see if we can modify the task in the queue vs. (re)scheduling the task in sync plugin.  Not necessarily optimistic but we'll see.

The thought of addressing this somehow at a servlet filter level also crossed my mind, but I ruled that out fairly quickly.  Bottom line, very hacky.

**Comment by bparees** [comment]
Let's figure out what the right priority is for this...we have a lot of other items on our backlog(not necessarily jenkins related). @csrwng can you help us understand how critical this is either for you, or what you think the impact to other users might be?

**Comment by csrwng** [comment]
So far it's not blocking anything. We've reverted to using regular Jenkins jobs which work ok. It is something that we should try to fix if we can though, since it prevents us from using our own pipeline build configs. I'm not sure how likely it is for a customer to hit this. If you want a rest api to start jobs you could use the OpenShift api to instantiate the buildconfig and that should work as expected. In my case, we are integrating with a different piece of software (Prow) which only knows how to talk to Jenkins and is expecting a particular response from Jenkins.

**Comment by gabemontero** [comment]
OK, as a disclaimer @bparees, I spent much more time sorting through the class loading issue we discussed on IRC than further experimentation here :-)  

And this exercise was educational wrt understanding elements of the sync plugin I previously did not have to look into.

I have a patch that enables the http RC 201 path.  It could use some more cleanup, and if we ever moved forward with this, my first thought is that this is not default behavior, but rather a behavior you explicitly enable.

Details:
1) getting the openshift build number and jenkins job/run number in sync appears problematic since sync plugin no longer schedules the run and primes it ... but maybe the cleanup I mentioned above has some bearing here.  I stopped coding when I got the basics working.

2) regardless of a pipeline strategy job, or a vanilla jenkins pipeline, when triggering from the console, in the http response location header associated with the 201, I'm seeing values like `https://jenkins-myproject.127.0.0.1.nip.io/queue/item/9/`  .... I'm thinking that is not what you were expecting, right @csrwng?  if such a location was unexpected even in the non-pipeline case, maybe that is something may patch somehow induced, could be corrected (though it is not  immediately apparent to me).

3) all the various metadata on the openshift build itself still looks good though wrt job urls, status json that openshift console uses for the pipeline stages/steps, etc.

4) I don't think we have to worry about about the git/scm trigger metadata that sync plugin adds, since this mode only applies to triggering the run from a curl / console interaction ... that stuff is also lost when sync plugin let's jenkins schedule the run.

So I'm thinking I could either save this and resume when time permits, or at least push a commit and let @csrwng tinker with it some more in the interim if he is so inclined.

Thoughts / opinions ?

**Comment by bparees** [comment]
> getting the openshift build number and jenkins job/run number in sync appears problematic since sync plugin no longer schedules the run and primes it

remind me what the plugin does currently?  does it set the jenkins job # to the openshift build number?

> I'm seeing values like https://jenkins-myproject.127.0.0.1.nip.io/queue/item/9/ .... I'm thinking that is not what you were expecting

what would be expected?  are you referring to the fact that it's linking into the queue instead of linking to the job itself?

**Comment by gabemontero** [comment]
On Fri, Sep 29, 2017 at 4:28 PM, Ben Parees <notifications@github.com>
wrote:

> getting the openshift build number and jenkins job/run number in sync
> appears problematic since sync plugin no longer schedules the run and
> primes it
>
> remind me what the plugin does currently? does it set the jenkins job # to
> the openshift build number?
>
I *think* that is it ... it creates a jenkins "Cause", where the key for
that cause is the build number ... I *think* jenkins uses that,
but am not 100% certain.  The disconnect I saw in testing could be either
an issue that exist regardless that I stumbled upon
with my triggers / restarts etc.

> I'm seeing values like https://jenkins-myproject.127.
> 0.0.1.nip.io/queue/item/9/ .... I'm thinking that is not what you were
> expecting
>
> what would be expected? are you referring to the fact that it's linking
> into the queue instead of linking to the job itself?
>
yes, link to the job itself;  I *thought* that is what @csrwng told me was
happening when talked face to face, but perhaps I misinterpreted;
certainly I could try with the current official version of the plugin and
compare, but decided to stop and report here before pursuing further

**Comment by gabemontero** [comment]
Also, based on the code I see related to the 201:

```
rsp.sendRedirect(SC_CREATED, req.getContextPath() + '/' + item.getUrl());
```

The queue item URL actually seems like what should be expected.

**Comment by gabemontero** [comment]
Sorry, a little more context... from `jenkins.model.ParameterizedJobMixIn<JobT extends Job<JobT, RunT> & ParameterizedJob & Task, RunT extends Run<JobT, RunT> & Executable>`:
```
        Queue.Item item = Jenkins.getInstance().getQueue().schedule2(asJob(), delay.getTime(), getBuildCause(asJob(), req)).getItem();
        if (item != null) {
            rsp.sendRedirect(SC_CREATED, req.getContextPath() + '/' + item.getUrl());
        } else {
            rsp.sendRedirect(".");
        }
```

**Comment by csrwng** [comment]
@gabemontero it is the queue URL that prow looks for when it invokes the build, sorry I got them mixed up:
https://github.com/kubernetes/test-infra/blob/master/prow/jenkins/controller.go#L302
https://github.com/kubernetes/test-infra/blob/master/prow/jenkins/controller.go#L310

it then uses that to figure out the state of the queued job:
https://github.com/kubernetes/test-infra/blob/master/prow/jenkins/jenkins.go#L155

**Comment by gabemontero** [comment]
On Fri, Sep 29, 2017 at 5:28 PM, Cesar Wong <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> it is the queue URL that
> prow looks for when it invokes the build, sorry I got them mixed up:
> https://github.com/kubernetes/test-infra/blob/master/prow/
> jenkins/controller.go#L302
> https://github.com/kubernetes/test-infra/blob/master/prow/
> jenkins/controller.go#L310
>
> it then uses that to figure out the state of the queued job:
> https://github.com/kubernetes/test-infra/blob/master/prow/
> jenkins/jenkins.go#L155
>
ok cool, so we are good with that piece; thanks for the update

**Comment by gabemontero** [comment]
OK, circling back to this for a bit.  Upon revisit, I realized I should explicitly point out that if we defer job scheduling to jenkins, we will lose env vars from the build and build config getting seeded into the job.

Unlike the build config Git related items (which would not pertain to job initiated from the console), the env var stuff is more of a sacrifice, and reinforces for me my initial thought that this form of sxecution should not be default behavior.

That said, going to do some more unit testing, revisit the build number / job number stuff a bit, and hopefully have something clean enough for @csrwng to try by EOB today.

**Comment by gabemontero** [comment]
was incorrect on the env vars .... the build config defaults are set up during job definition

they should be available

**Comment by gabemontero** [comment]
btw @bparees after about 15 or so plugin reloads I finally hit the class loading issue:

```
SEVERE: Timer task io.fabric8.jenkins.openshiftsync.ConfigMapWatcher$1@13f1eed failed
java.lang.LinkageError: loader constraint violation: when resolving method "com.fasterxml.jackson.databind.ObjectMapper.<init>(Lcom/fasterxml/jackson/core/JsonFactory;)V" the class loader (instance of hudson/PluginFirstClassLoader) of the current class, io/fabric8/kubernetes/client/utils/Serialization, and the class loader (instance of hudson/PluginFirstClassLoader) for the method's defining class, com/fasterxml/jackson/databind/ObjectMapper, have different Class objects for the type com/fasterxml/jackson/core/JsonFactory used in the signature
	at io.fabric8.kubernetes.client.utils.Serialization.<clinit>(Serialization.java:38)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.<clinit>(OperationSupport.java:59)
	at io.fabric8.kubernetes.client.DefaultKubernetesClient.configMaps(DefaultKubernetesClient.java:254)
	at io.fabric8.openshift.client.DefaultOpenShiftClient.configMaps(DefaultOpenShiftClient.java:302)
	at io.fabric8.jenkins.openshiftsync.ConfigMapWatcher$1.doRun(ConfigMapWatcher.java:66)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:51)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by gabemontero** [comment]
ok .. i've got an acceptable prototype for @csrwng to try; I'm now much more comfortable that this *could* be the default mode of operation when initiating pipeline strategy runs via the jenkins console or curl posts to jenkins; I believe concerns I have noted previously are either resolved, or whose deferring can be rationalized

I had to get a little creative on the jenkins run to openshift build correlation (the difficulties of this are certainly when the existing implementation went the way it did).  But it looks like it should hold up, be thread safe etc.

We can get into those particulars when I create a PR for this (obviously a post 3.7 items).

But for now, give 
[sync-plugin.zip](https://github.com/openshift/jenkins-sync-plugin/files/1356741/sync-plugin.zip)
a try @csrwng and let me know how it goes.

thanks

**Comment by gabemontero** [comment]
and no linkage errors since i bumped pipeline-rest-api yesterday ... multiple plugin redeploys

**Comment by csrwng** [comment]
@gabemontero I tested your fix above and so far it's been working great for me, even when starting multiple jobs simultaneously.

**Comment by csrwng** [comment]
actually, not quite (sorry, spoke too soon :-) ). So when starting multiple jobs, the jobs work great in Jenkins. However for whatever reason they're not properly synced with the openshift build objects. In Jenkins, all builds finished successfully. However, on the openshift side some jobs are in pending, others are in new.

**Comment by gabemontero** [comment]
ah, *almost* great news @csrwng ; thanks for trying and reporting.  That was an area where I had to iterate a bit to get it working even for me.

If you can, capture / send the existing Jenkins logs, and we'll see if there is an obvious exception.

Otherwise, If you could get a repro with *ALL* level logging on for these classes:
- io.fabric8.jenkins.openshiftsync.BuildWatcher
- io.fabric8.jenkins.openshiftsync.OpenShiftUtils
- io.fabric8.jenkins.openshiftsync.JenkinsUtils

That could shed even more light.

If still no luck, I'll have to craft a debug jar.

thanks

**Comment by gabemontero** [comment]
@bparees - this item of initiating a pipeline build run via a curl against the rest endpoint slipped through the cracks without a trello card

Shall I open a new card on the devtools candidates board, or wait until Jira comes around?

**Comment by bparees** [comment]
@gabemontero you can create a trello card if you put it directly into the devtools candidates list.

**Comment by gabemontero** [comment]
Created https://trello.com/c/sckb6xVi/1588-5-trigger-job-via-direct-rest-invocation to track this

**Comment by cglez1987** [comment]
Hi @gabemontero. I wonder if this issue was solved cause we are having the same problem as @csrwng. Please this is a high priority for us. Thanks in advance.

**Comment by gabemontero** [comment]
> Hi @gabemontero. I wonder if this issue was solved cause we are having the same problem as @csrwng. Please this is a high priority for us. Thanks in advance.

hi @cglez1987 

first, there are multiple issues discussed in this relatively ancient issue :-)

In general, I *think* the curl/post/rc stuff was sorted out.  I've heard no discussion about this in quite a while.  I thought we got it working with work around that trello item, though we don't invoke jenkins from out internal prow anymore, so unless we have other customers using that, not sure how often that flavor of invocation is used. Unfortunately, we no longer use trello and I cannot access that trello card to see what the outcome was.

If you are referring to https://github.com/openshift/jenkins-sync-plugin/issues/175#issuecomment-335164289 and items stuck in pending, etc. it turns out that was an intermittent timing bug that has only been fixed rather recently via https://github.com/openshift/jenkins-sync-plugin/pull/412 and is in v1.0.53 of the plugin.

If it is that, ^^, great, otherwise, you would be best served opening a support case with Red Hat, supplying your various version information, reproducing `curl` invocations if that is relevant, where the team that currently supports openshift jenkins can triage (I am transitioning to another project so I may not able to provide you significantly more cycles on this if you need them).

If you don't pay for RH support, I would suggest opening a new issue on this repo with all the pertinent info and things can get processed on a best effort basis.

@coreydaley @akram FYI

**Comment by cglez1987** [comment]
Thanks @gabemontero for your explanation. Unfortunately,  we are facing the issue with curl command returning 302 instead of 201. We are going to review again our versions and if we have a major version of the plugin, we'll open a support case with RH. Thanks again.

**Comment by gabemontero** [comment]
> Thanks @gabemontero for your explanation. Unfortunately, we are facing the issue with curl command returning 302 instead of 201. We are going to review again our versions and if we have a major version of the plugin, we'll open a support case with RH. Thanks again.

sounds good

definitely include
- ocp version (including x.x.x granularity)
- jenkins version, in case you have upgraded jenkins after your last OCP upgrade)
- sync plugin version, in case you have upgraded after your last OCP upgrade)
- as much of the curl invocation as you can supply, as much of the rc processing as you can supply
- are you in fact using the jenkins pipeline build strategy

that last point may be key; we'll have to refresh our collective memories on jenkins behavior both when sync plugin is involved and when it is not involved; if this is in fact a core jenkins issue, we won't be able to do much to help you, other than maybe some research on upstream issues around this.... I don't remermber where my activity 4 years ago landed with respect to that.

That last point ^^ might be key as

**Comment by cglez1987** [comment]
Hi @gabemontero. Really appreciate your dedication to this. We are going to open a ticket/support for this problem and in the meantime, we'll try to figure out a workaround possibly using that command. Thanks again and I'll keep you posted.
